File: src/main/scala/com/michalplachta/galactic/internal/PredicateLogic.scala
package com.michalplachta.galactic.internal

object PredicateLogic {
  implicit class PredicateWithLogic[A](p1: A ⇒ Boolean) {
    def and(p2: A ⇒ Boolean): A ⇒ Boolean = {
      a ⇒ p1(a) && p2(a)
    }

    def andNot(p2: A ⇒ Boolean): A ⇒ Boolean = {
      a ⇒ p1(a) && !p2(a)
    }
  }

  def not[A](p: A ⇒ Boolean): A ⇒ Boolean = a ⇒ !p(a)

  def always[A]: A ⇒ Boolean = _ ⇒ true
}

File: src/main/java/com/michalplachta/galactic/java/internal/PredicateLogic.java
package com.michalplachta.galactic.java.internal;

import io.vavr.Function1;

public class PredicateLogic {
    public static <T> Function1<T, Boolean> and(Function1<T, Boolean> f1, Function1<T, Boolean> f2) {
        return Function1.of(t -> f1.apply(t) && f2.apply(t));
    }

    public static <T> Function1<T, Boolean> not(Function1<T, Boolean> f) {
        return Function1.of(t -> !f.apply(t));
    }

    public static <T> Function1<T, Boolean> always() {
        return Function1.of(t -> true);
    }
}
